In [ ]:
from google.colab import drive
drive.mount('/content/drive/')
%cd /content/drive/My Drive/DTSC2301


import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression



Drive already mounted at /content/drive/; to attempt to forcibly remount, call drive.mount("/content/drive/", force_remount=True).
/content/drive/My Drive/DTSC2301


In [ ]:
df = pd.read_csv("energy_dataset.csv (1).txt",parse_dates=['time'],index_col=['time'])



#DataFrameGroupBy = pd.read_csv("energy_dataset.csv (1).txt",parse_dates=['time'],index_col=['time'])

df.index = pd.to_datetime(df.index, utc='True')

#df.columns =['time','generation biomass','generation fossil brown coal/lignite','generation fossil coal-derived gas','generation fossil gas','generation fossil hard coal','generation fossil oil','generation fossil oil shale','generation fossil peat','generation geothermal','generation hydro pumped storage aggregated','generation hydro pumped storage consumption','generation hydro run-of-river and poundage','generation hydro water reservoir','generation marine','generation nuclear','generation other','generation other renewable','generation solar','generation waste','generation wind offshore','generation wind onshore','forecast solar day ahead','forecast wind offshore eday ahead','forecast wind onshore day ahead','total load forecast','total load actual','price day ahead','price actual']
#df.set_index('time',inplace = True)
df


generation biomass  \
time                                            
2014-12-31 23:00:00+00:00               447.0   
2015-01-01 00:00:00+00:00               449.0   
2015-01-01 01:00:00+00:00               448.0   
2015-01-01 02:00:00+00:00               438.0   
2015-01-01 03:00:00+00:00               428.0   
...                                       ...   
2018-12-31 18:00:00+00:00               297.0   
2018-12-31 19:00:00+00:00               296.0   
2018-12-31 20:00:00+00:00               292.0   
2018-12-31 21:00:00+00:00               293.0   
2018-12-31 22:00:00+00:00               290.0   

                           generation fossil brown coal/lignite  \
time                                                              
2014-12-31 23:00:00+00:00                                 329.0   
2015-01-01 00:00:00+00:00                                 328.0   
2015-01-01 01:00:00+00:00                                 323.0   
2015-01-01 02:00:00+00:00                                 254.0   
2015-01-01 03:00:00+00:00                                 187.0   
...                                                         ...   
2018-12-31 18:00:00+00:00                                   0.0   
2018-12-31 19:00:00+00:00                                   0.0   
2018-12-31 20:00:00+00:00                                   0.0   
2018-12-31 21:00:00+00:00                                   0.0   
2018-12-31 22:00:00+00:00                                   0.0   

                           generation fossil coal-derived gas  \
time                                                            
2014-12-31 23:00:00+00:00                                 0.0   
2015-01-01 00:00:00+00:00                                 0.0   
2015-01-01 01:00:00+00:00                                 0.0   
2015-01-01 02:00:00+00:00                                 0.0   
2015-01-01 03:00:00+00:00                                 0.0   
...                                                       ...   
2018-12-31 18:00:00+00:00                                 0.0   
2018-12-31 19:00:00+00:00                                 0.0   
2018-12-31 20:00:00+00:00                                 0.0   
2018-12-31 21:00:00+00:00                                 0.0   
2018-12-31 22:00:00+00:00                                 0.0   

                           generation fossil gas  generation fossil hard coal  \
time                                                                            
2014-12-31 23:00:00+00:00                 4844.0                       4821.0   
2015-01-01 00:00:00+00:00                 5196.0                       4755.0   
2015-01-01 01:00:00+00:00                 4857.0                       4581.0   
2015-01-01 02:00:00+00:00                 4314.0                       4131.0   
2015-01-01 03:00:00+00:00                 4130.0                       3840.0   
...                                          ...                          ...   
2018-12-31 18:00:00+00:00                 7634.0                       2628.0   
2018-12-31 19:00:00+00:00                 7241.0                       2566.0   
2018-12-31 20:00:00+00:00                 7025.0                       2422.0   
2018-12-31 21:00:00+00:00                 6562.0                       2293.0   
2018-12-31 22:00:00+00:00                 6926.0                       2166.0   

                           generation fossil oil  generation fossil oil shale  \
time                                                                            
2014-12-31 23:00:00+00:00                  162.0                          0.0   
2015-01-01 00:00:00+00:00                  158.0                          0.0   
2015-01-01 01:00:00+00:00                  157.0                          0.0   
2015-01-01 02:00:00+00:00                  160.0                          0.0   
2015-01-01 03:00:00+00:00                  156.0                          0.0   
...                                          ...  

In [ ]:
#check for null and missing values
df.isna().sum()


generation biomass                                19
generation fossil brown coal/lignite              18
generation fossil coal-derived gas                18
generation fossil gas                             18
generation fossil hard coal                       18
generation fossil oil                             19
generation fossil oil shale                       18
generation fossil peat                            18
generation geothermal                             18
generation hydro pumped storage aggregated     35064
generation hydro pumped storage consumption       19
generation hydro run-of-river and poundage        19
generation hydro water reservoir                  18
generation marine                                 19
generation nuclear                                17
generation other                                  18
generation other renewable                        18
generation solar                                  18
generation waste                              

In [ ]:
df.isnull().sum()

generation biomass                                19
generation fossil brown coal/lignite              18
generation fossil coal-derived gas                18
generation fossil gas                             18
generation fossil hard coal                       18
generation fossil oil                             19
generation fossil oil shale                       18
generation fossil peat                            18
generation geothermal                             18
generation hydro pumped storage aggregated     35064
generation hydro pumped storage consumption       19
generation hydro run-of-river and poundage        19
generation hydro water reservoir                  18
generation marine                                 19
generation nuclear                                17
generation other                                  18
generation other renewable                        18
generation solar                                  18
generation waste                              

In [ ]:
#so far, there are 7 columns that need to be removed, as they do not provide any useful data


#generation hydro pumped storage aggregated     35064
#forecast wind offshore eday ahead              35064


df.sum()


generation biomass                             1.344023e+07
generation fossil brown coal/lignite           1.570268e+07
generation fossil coal-derived gas             0.000000e+00
generation fossil gas                          1.970545e+08
generation fossil hard coal                    1.491581e+08
generation fossil oil                          1.045462e+07
generation fossil oil shale                    0.000000e+00
generation fossil peat                         0.000000e+00
generation geothermal                          0.000000e+00
generation hydro pumped storage aggregated     0.000000e+00
generation hydro pumped storage consumption    1.666661e+07
generation hydro run-of-river and poundage     3.406781e+07
generation hydro water reservoir               9.129885e+07
generation marine                              0.000000e+00
generation nuclear                             2.195312e+08
generation other                               2.110771e+06
generation other renewable              

In [ ]:
#we now should count our 0's along with the rest of them, as they could affect the mean. the plan here, is to later impute the missing values in the dataset.

for column_name in df.columns:

  zerodf = (df[column_name] == 0).sum()
  print(f"{column_name:{50}} : {zerodf}")




generation biomass                                 : 4
generation fossil brown coal/lignite               : 10517
generation fossil coal-derived gas                 : 35046
generation fossil gas                              : 1
generation fossil hard coal                        : 3
generation fossil oil                              : 3
generation fossil oil shale                        : 35046
generation fossil peat                             : 35046
generation geothermal                              : 35046
generation hydro pumped storage aggregated         : 0
generation hydro pumped storage consumption        : 12607
generation hydro run-of-river and poundage         : 3
generation hydro water reservoir                   : 3
generation marine                                  : 35045
generation nuclear                                 : 3
generation other                                   : 4
generation other renewable                         : 3
generation solar                     

In [ ]:
#clean data by dropping unuseful columns, and imputing missing values
#https://www.statista.com/statistics/737763/energy-production-from-renewable-municipal-waste-in-spain/
#We decided that including some zeros and imputing others (like some random zeros for things we can judge as a missing value)

#df.drop(['generation hydro pumped storage aggregated', 'forecast wind offshore eday ahead',
        #'generation wind offshore', 'generation fossil coal-derived gas',
      #  'generation fossil oil shale', 'generation fossil peat', 'generation marine',
      #     'generation wind offshore', 'generation geothermal'], inplace = True, axis = 1)

df.dropna()



Empty DataFrame
Columns: [generation biomass, generation fossil brown coal/lignite, generation fossil coal-derived gas, generation fossil gas, generation fossil hard coal, generation fossil oil, generation fossil oil shale, generation fossil peat, generation geothermal, generation hydro pumped storage aggregated, generation hydro pumped storage consumption, generation hydro run-of-river and poundage, generation hydro water reservoir, generation marine, generation nuclear, generation other, generation other renewable, generation solar, generation waste, generation wind offshore, generation wind onshore, forecast solar day ahead, forecast wind offshore eday ahead, forecast wind onshore day ahead, total load forecast, total load actual, price day ahead, price actual]
Index: []

[0 rows x 28 columns]

In [ ]:
df.describe()

generation biomass  generation fossil brown coal/lignite  \
count        35045.000000                          35046.000000   
mean           383.513540                            448.059208   
std             85.353943                            354.568590   
min              0.000000                              0.000000   
25%            333.000000                              0.000000   
50%            367.000000                            509.000000   
75%            433.000000                            757.000000   
max            592.000000                            999.000000   

       generation fossil coal-derived gas  generation fossil gas  \
count                             35046.0           35046.000000   
mean                                  0.0            5622.737488   
std                                   0.0            2201.830478   
min                                   0.0               0.000000   
25%                                   0.0            4126.000000   
50%                                   0.0            4969.000000   
75%                                   0.0            6429.000000   
max                                   0.0           20034.000000   

       generation fossil hard coal  generation fossil oil  \
count                 35046.000000           35045.000000   
mean                   4256.065742             298.319789   
std                    1961.601013              52.520673   
min                       0.000000               0.000000   
25%                    2527.000000             263.000000   
50%                    4474.000000             300.000000   
75%                    5838.750000             330.000000   
max                    8359.000000             449.000000   

       generation fossil oil shale  generation fossil peat  \
count                      35046.0                 35046.0   
mean                           0.0                     0.0   
std                            0.0                     0.0   
min                            0.0                     0.0   
25%                            0.0                     0.0   
50%                            0.0                     0.0   
75%                            0.0                     0.0   
max                            0.0                     0.0   

       generation geothermal  generation hydro pumped storage aggregated  ...  \
count                35046.0                                         0.0  ...   
mean                     0.0                                         NaN  ...   
std                      0.0                                         NaN  ...   
min                      0.0                                         NaN  ...   
25%                      0.0                                         NaN  ...   
50%                      0.0                                         NaN  ...   
75%                      0.0                                         NaN  ...   
max                      0.0                                         NaN  ...   

       generation waste  generation wind offshore  generation wind onshore  \
count      35045.000000                   35046.0             35046.000000   
mean         269.452133                       0.0              5464.479769   
std           50.195536                       0.0              3213.691587   
min            0.000000                       0.0                 0.000000   
25%          240.000000                       0.0              2933.000000   
50%          279.000000                       0.0              4849.000000   
75%          310.000000                       0.0              7398.000000   
max          357.000000                       0.0             17436.000000   

       forecast solar day ahead  forecast wind offshore eday ahead  \
count              35064.000000                                0.0   
mean                1439.066735                                NaN   
std                 1677.703355       

generation biomass                                 : 4
generation fossil brown coal/lignite               : 10517
generation fossil coal-derived gas                 : 35046
generation fossil gas                              : 1
generation fossil hard coal                        : 3
generation fossil oil                              : 3
generation fossil oil shale                        : 35046
generation fossil peat                             : 35046
generation geothermal                              : 35046
generation hydro pumped storage aggregated         : 0
generation hydro pumped storage consumption        : 12607
generation hydro run-of-river and poundage         : 3
generation hydro water reservoir                   : 3
generation marine                                  : 35045
generation nuclear                                 : 3
generation other                                   : 4
generation other renewable                         : 3
generation solar                                   : 3
generation waste                                   : 3
generation wind offshore                           : 35046
generation wind onshore                            : 3
forecast solar day ahead                           : 539
forecast wind offshore eday ahead                  : 0
forecast wind onshore day ahead                    : 0
total load forecast                                : 0
total load actual                                  : 0
price day ahead                                    : 0
price actual        

#Regression 19 features

In [ ]:
#fit a model for desciion tree

from sklearn.tree import DecisionTreeClassifier, DecisionTreeRegressor

from sklearn import tree
from sklearn.model_selection import train_test_split
import matplotlib.pyplot as plt
from sklearn.metrics import mean_squared_error,r2_score
from sklearn.impute import SimpleImputer
from sklearn import tree
import matplotlib.pyplot as plt
import statsmodels.api as sm
from sklearn.tree import export_graphviz
import graphviz

#impute missing values
imputer = SimpleImputer(strategy='mean')
#bc data is all numeric values, it is unnesscary to convert to dummies, so we just define our data
newdf = pd.DataFrame(imputer.fit_transform(df), columns=df.columns)
X = newdf.drop('price actual', axis=1)

y = newdf['price actual']

#create instance of model
regressor = DecisionTreeRegressor()

#randomize and evenly split our data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


#train the data we've just split
X_train_imputed = imputer.fit_transform(X_train)
X_test_imputed = imputer.transform(X_test)

#fit the model to trained data
regressor.fit(X_train_imputed, y_train)


#predict
y_pred = regressor.predict(X_test_imputed)


#model evaluation
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)
print('Mean Squared Error:', mse)
print('R^2 Score:', r2)

#coloring to maintain repot constostency

#dot_data = export_graphviz(regressor, out_file= None, feature_names= X.columns,  filled=True, rounded=True,  special_characters=True)


#graph = graphviz.Source(dot_data)

#graph.format = 'png'  # Output format (e.g., png, pdf, svg, etc.)

#graph.render('decision_tree')

#plot the fig
plt.figure(figsize=(20,15))
tree.plot_tree(regressor,fontsize = 9, filled=True, feature_names=newdf.columns,max_depth=3, label="root",rounded=True)
plt.show()



ValueError: Shape of passed values is (35064, 26), indices imply (35064, 28)

In [ ]:
print(newdf.columns)

 A decisioon tree provides extremely useful insight into which variables our model is best suited to predict, and gives us the opportunity to expolore them further.This model is also very useful in the context of our investigation, as descicion trees are highly capable of handling missing values while maintaining the integrity of the data.

 It also reassures us that the average price day ahead is indeed a stronge indicator of the actual price, which aligns with our percection of the context in the data.

 MITECO - Ministry for the Ecological Transition
and the Demographical Challenge
(www.miteco.gob.es/es)

In [ ]:
#twelevedf= newdf.drop([ 'generation hydro pumped storage consumption','generation fossil brown coal/lignite'],inplace=True, axis = 1 )

In [ ]:
#summary of model capacity


rmse = np.sqrt(mse)

print(rmse)


In [ ]:
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score

mse = mean_squared_error(y_test, y_pred)
print(f"Mean Squared Error (MSE): {mse}")

# Root Mean Squared Error (RMSE)
rmse = np.sqrt(mse)
print(f"Root Mean Squared Error (RMSE): {rmse}")

# Mean Absolute Error (MAE)
mae = mean_absolute_error(y_test, y_pred)
print(f"Mean Absolute Error (MAE): {mae}")

# R-squared (R²)
r2 = r2_score(y_test, y_pred)
print(f"R-squared (R²): {r2}")


In [ ]:
#adjusred R2

n = len(y)
k = X.shape[1]

adjusted_r2 = 1 - (1 - r2) * (n - 1) / (n - k - 1)

print("R-squared:", r2)
print("Adjusted R-squared:", adjusted_r2)

# Residuals for 20 feature

In [ ]:
residuals = y_test - y_pred
plt.figure(figsize=(10, 6))
plt.scatter(y_pred, residuals, alpha=0.5)
plt.axhline(y=0, color='r', linestyle='--')  # Adds a horizontal line at y=0
plt.xlabel('Predicted')
plt.ylabel('Residuals')
plt.title('Residuals of Predictions')
plt.show()


#Linear regression for 20 feature

In [ ]:
#linear regression model
Xlinear = newdf.drop('price actual', axis=1)
ylinear = newdf['price actual']

Xlin_train, Xlin_test, ylin_train, ylin_test = train_test_split(Xlinear, ylinear, test_size=0.3, random_state=42)

#add constant
Xlin_train = sm.add_constant(Xlin_train)
Xlin_test = sm.add_constant(Xlin_test)
# Impute missing values with mean
imputer = SimpleImputer(strategy='mean')


# Fit linear regression model
model = LinearRegression()

model.fit(Xlin_train, ylin_train)

# Predict on test set
ylin_pred = model.predict(Xlin_train)

# Plot actual vs predicted
# Scatter plot of actual vs. predicted values
plt.figure(figsize=(10, 6))
plt.scatter(ylin_test, ylin_pred, alpha=0.5)
plt.plot([ylin_test.min(), ylin_test.max()], [ylin_test.min(), ylin_test.max()], '--k')
plt.xlabel('Actual Price')
plt.ylabel('Predicted Price')
plt.title('Actual Price vs. Predicted Price')
plt.show()




In [ ]:
import statsmodels.api as sm


lm1 = sm.OLS(ylin_train, Xlin_train).fit()
#Print Model Summary
print(lm1.summary())

In [ ]:
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score

mse = mean_squared_error(ylin_test, ylin_pred)
print(f"Mean Squared Error (MSE): {mse}")

# Root Mean Squared Error (RMSE)
rmse = np.sqrt(mse)
print(f"Root Mean Squared Error (RMSE): {rmse}")

# Mean Absolute Error (MAE)
mae = mean_absolute_error(ylin_test, ylin_pred)
print(f"Mean Absolute Error (MAE): {mae}")

# R-squared (R²)
r2 = r2_score(ylin_test, ylin_pred)
print(f"R-squared (R²): {r2}")

